core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala

```diff
@@ -565,12 +603,25 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
   }
 
   /**
-   * Simplified version of combineByKey that hash-partitions the resulting RDD using the
-   * existing partitioner/parallelism level.
+   * This method is here for backward compatibility. It
+   * does not provide combiner classtag information to
+   * the shuffle.
+   *
+   * @see [[combineByKeyWithClassTag]]
    */
   def combineByKey[C](createCombiner: V => C, mergeValue: (C, V) => C, mergeCombiners: (C, C) => C)
-    : RDD[(K, C)] = self.withScope {
-    combineByKey(createCombiner, mergeValue, mergeCombiners, defaultPartitioner(self))
+    : RDD[(K, C)] = {
+    combineByKeyWithClassTag(createCombiner, mergeValue, mergeCombiners)(null)
+  }
+
+  /**
+   * Simplified version of combineByKeyWithClassTag that hash-partitions the resulting RDD using the
+   * existing partitioner/parallelism level.
+   */
+  def combineByKeyWithClassTag[C](createCombiner: V => C, mergeValue: (C, V) => C,
```

Comment:
Adding a `ClassTag` to the original `combineByKey` methods, would change the method signature break compatibility. The `PairRDDFunctions` is not marked a `@DeveloperApi`, so that break is not allowed.

To work around this, there are now three new methods that exactly mirror the `combineByKey` methods but have a `*withClassTag` suffix on their name and, of course, a `ClassTag`. We want developers to use the `combineByKeyWithClassTag` methods need to be `public` just as the original `combineByKey` methods are.